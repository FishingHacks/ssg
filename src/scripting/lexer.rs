use std::collections::VecDeque;
use std::ops::ControlFlow;
use std::sync::Arc;

use super::{ByteOffset, ParsingError};

#[derive(Debug)]
enum LexingMode {
    Html,
    Script,
}

#[derive(Debug, Clone, Copy)]
pub enum Token {
    Html(ByteOffset),
    ScriptStart(ByteOffset),
    ScriptEnd(ByteOffset),
    Identifier(ByteOffset),
    StringLiteral(ByteOffset),
    Float(ByteOffset),
    Int(ByteOffset),

    Assign(ByteOffset),
    Equal(ByteOffset),
    NotEqual(ByteOffset),
    GreaterEqual(ByteOffset),
    LesserEqual(ByteOffset),
    Greater(ByteOffset),
    Lesser(ByteOffset),
    Dot(ByteOffset),
    LeftParen(ByteOffset),
    RightParen(ByteOffset),
    Minus(ByteOffset),
    Plus(ByteOffset),
    Mul(ByteOffset),

    Comma(ByteOffset),

    For(ByteOffset),
    In(ByteOffset),
    End(ByteOffset),
    If(ByteOffset),
    Else(ByteOffset),
    Block(ByteOffset),
    Render(ByteOffset),
    Slot(ByteOffset),
    Extend(ByteOffset),
}

impl Token {
    pub fn loc(&self) -> ByteOffset {
        match self {
            Token::Html(byte_offset)
            | Token::ScriptStart(byte_offset)
            | Token::ScriptEnd(byte_offset)
            | Token::Identifier(byte_offset)
            | Token::StringLiteral(byte_offset)
            | Token::Float(byte_offset)
            | Token::Assign(byte_offset)
            | Token::Equal(byte_offset)
            | Token::NotEqual(byte_offset)
            | Token::GreaterEqual(byte_offset)
            | Token::LesserEqual(byte_offset)
            | Token::Greater(byte_offset)
            | Token::Lesser(byte_offset)
            | Token::Dot(byte_offset)
            | Token::LeftParen(byte_offset)
            | Token::RightParen(byte_offset)
            | Token::Minus(byte_offset)
            | Token::Plus(byte_offset)
            | Token::For(byte_offset)
            | Token::In(byte_offset)
            | Token::End(byte_offset)
            | Token::If(byte_offset)
            | Token::Else(byte_offset)
            | Token::Block(byte_offset)
            | Token::Comma(byte_offset)
            | Token::Render(byte_offset)
            | Token::Slot(byte_offset)
            | Token::Mul(byte_offset)
            | Token::Extend(byte_offset)
            | Token::Int(byte_offset) => *byte_offset,
        }
    }

    pub fn is_operator(&self) -> bool {
        matches!(
            self,
            Token::Plus(_)
                | Token::Minus(_)
                | Token::LeftParen(_)
                | Token::RightParen(_)
                | Token::Lesser(_)
                | Token::LesserEqual(_)
                | Token::Greater(_)
                | Token::GreaterEqual(_)
                | Token::NotEqual(_)
                | Token::Equal(_)
                | Token::Dot(_)
                | Token::Mul(_)
        )
    }

    pub fn is_keyword(&self) -> bool {
        matches!(
            self,
            Token::For(_)
                | Token::In(_)
                | Token::If(_)
                | Token::Render(_)
                | Token::Extend(_)
                | Token::Block(_)
                | Token::Slot(_)
                | Token::Else(_)
        )
    }
}

#[derive(Debug)]
pub struct Lexer {
    cursor: usize,
    pub source: Arc<String>,
    mode: LexingMode,
    html_start: Option<usize>,
    script_start: Option<usize>,
    peeked: VecDeque<Result<Token, ParsingError>>,
}

impl Lexer {
    pub fn new(source: Arc<String>) -> Self {
        Self {
            source,
            cursor: 0,
            html_start: None,
            script_start: None,
            peeked: VecDeque::new(),
            mode: LexingMode::Html,
        }
    }

    pub fn peek_n(&mut self, n: usize) -> Option<Result<Token, ParsingError>> {
        let mut buffer = Vec::with_capacity(n);

        while self.peeked.len() <= n {
            let Some(next) = self.next() else {
                break;
            };

            buffer.push(next);
        }

        self.peeked.extend(buffer);

        self.peeked.get(n).cloned()
    }

    pub fn peek(&mut self) -> Option<Result<Token, ParsingError>> {
        self.peek_n(0)
    }

    pub fn is_empty(&mut self) -> bool {
        self.peek().is_none()
    }

    fn is_keyword(&self, ident: &str, offset: ByteOffset) -> Option<Token> {
        match ident {
            "for" => Some(Token::For(offset)),
            "in" => Some(Token::In(offset)),
            "end" => Some(Token::End(offset)),
            "if" => Some(Token::If(offset)),
            "else" => Some(Token::Else(offset)),
            "block" => Some(Token::Block(offset)),
            "render" => Some(Token::Render(offset)),
            "slot" => Some(Token::Slot(offset)),
            "extend" => Some(Token::Extend(offset)),
            _ => None,
        }
    }

    fn advance_script(&mut self, amount: usize) {
        self.cursor += amount;
        self.script_start = Some(self.cursor);
    }

    fn lex_html(&mut self, curr: char, next: Option<char>) -> ControlFlow<Result<Token, ParsingError>, ()> {
        match (curr, next) {
            ('{', Some('{')) => {
                self.mode = LexingMode::Script;
                self.script_start = Some(self.cursor);

                if let Some(start) = self.html_start.take() {
                    return ControlFlow::Break(Ok(Token::Html((start, self.cursor).into())));
                };

                ControlFlow::Continue(())
            }
            // reached the last character, so we return the whole section since
            // next char would be EOF
            (_, None) => {
                self.cursor += 1;
                let start = self.html_start.take().unwrap_or(self.cursor - 1);
                ControlFlow::Break(Ok(Token::Html((start, self.cursor).into())))
            }
            _ => {
                if self.html_start.is_none() {
                    self.html_start = Some(self.cursor);
                }

                self.cursor += 1;

                ControlFlow::Continue(())
            }
        }
    }

    fn lex_script(&mut self, curr: char, next: Option<char>) -> ControlFlow<Result<Token, ParsingError>> {
        match (curr, next) {
            (c, _) if c.is_whitespace() => {
                self.advance_script(c.len_utf8());
                ControlFlow::Continue(())
            }
            ('{', Some('{')) => {
                let start = self
                    .script_start
                    .expect("script start must be set while lexing script content");

                self.advance_script(2);

                ControlFlow::Break(Ok(Token::ScriptStart((start, self.cursor).into())))
            }
            ('}', Some('}')) => {
                self.mode = LexingMode::Html;
                self.cursor += 2;

                let start = self
                    .script_start
                    .take()
                    .expect("script start must be set while lexing script content");

                ControlFlow::Break(Ok(Token::ScriptEnd((start, self.cursor).into())))
            }
            (':', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::Assign((self.cursor - 2, self.cursor).into())))
            }
            ('=', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::Equal((self.cursor - 2, self.cursor).into())))
            }
            ('!', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::NotEqual((self.cursor - 2, self.cursor).into())))
            }
            ('>', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::GreaterEqual((self.cursor - 2, self.cursor).into())))
            }
            ('<', Some('=')) => {
                self.advance_script(2);
                ControlFlow::Break(Ok(Token::LesserEqual((self.cursor - 2, self.cursor).into())))
            }
            ('>', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Greater((self.cursor - 1, self.cursor).into())))
            }
            ('<', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Lesser((self.cursor - 1, self.cursor).into())))
            }
            ('.', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Dot((self.cursor - 1, self.cursor).into())))
            }
            ('(', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::LeftParen((self.cursor - 1, self.cursor).into())))
            }
            (')', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::RightParen((self.cursor - 1, self.cursor).into())))
            }
            ('-', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Minus((self.cursor - 1, self.cursor).into())))
            }
            ('*', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Mul((self.cursor - 1, self.cursor).into())))
            }
            (',', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Comma((self.cursor - 1, self.cursor).into())))
            }
            ('+', _) => {
                self.advance_script(1);
                ControlFlow::Break(Ok(Token::Plus((self.cursor - 1, self.cursor).into())))
            }
            ('"' | '\'', _) => {
                self.cursor += 1;
                let start = self.cursor;
                let source = &self.source[self.cursor..];

                let Some(length) = source.find(curr) else {
                    return ControlFlow::Break(Err(ParsingError {
                        src: self.source.clone(),
                        at: (start..self.source.len()).into(),
                        help: "unterminated string".into(),
                    }));
                };

                // skip past the closing " for the next iterations but don't
                // include it on the offset range
                self.advance_script(length + 1);
                let end = start + length;

                ControlFlow::Break(Ok(Token::StringLiteral((start, end).into())))
            }
            ('0'..='9', _) => {
                let start = self.cursor;
                let source = &self.source[self.cursor..];

                let length = source
                    .find(|c| !matches!(c, '_' | '.' | '0'..='9'))
                    .unwrap_or(self.source.len());

                self.advance_script(length);
                let end = start + length;
                let literal = &self.source[start..end];

                let mut dotted = literal.splitn(3, '.');
                match (dotted.next(), dotted.next(), dotted.next()) {
                    // found at least 2 dots on a number. thats a invalid float
                    (Some(_), Some(_), Some(_)) => {
                        return ControlFlow::Break(Err(ParsingError {
                            src: self.source.clone(),
                            at: (start..end).into(),
                            help: "invalid numeral".into(),
                        }));
                    }
                    _ => {
                        // literal is either a integer or a valid floating point
                        // so skip doing anything here
                    }
                };

                let is_float = literal.contains('.');
                if is_float {
                    return ControlFlow::Break(Ok(Token::Float((start, end).into())));
                }

                ControlFlow::Break(Ok(Token::Int((start, end).into())))
            }
            (c, _) => {
                if c.is_identifier_starter() {
                    let start = self.cursor;
                    let source = &self.source[self.cursor..];

                    let length = source
                        .find(|c: char| !c.is_valid_identifier())
                        .unwrap_or(self.source.len());

                    self.advance_script(length);

                    let end = start + length;
                    let content = &self.source[start..end];
                    let offset = (start, end).into();

                    if let Some(keyword) = self.is_keyword(content, offset) {
                        return ControlFlow::Break(Ok(keyword));
                    }

                    return ControlFlow::Break(Ok(Token::Identifier(offset)));
                }

                println!("{c:?}");
                panic!();
            }
        }
    }
}

impl Iterator for Lexer {
    type Item = Result<Token, ParsingError>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if let Some(peeked) = self.peeked.pop_front() {
                return Some(peeked);
            }

            let source = &self.source[self.cursor..];
            let mut chars = source.chars().peekable();

            let ch = chars.next()?;
            let next = chars.peek().copied();

            match self.mode {
                LexingMode::Html => match self.lex_html(ch, next) {
                    ControlFlow::Continue(_) => continue,
                    ControlFlow::Break(token) => return Some(token),
                },
                LexingMode::Script => match self.lex_script(ch, next) {
                    ControlFlow::Continue(_) => continue,
                    ControlFlow::Break(token) => return Some(token),
                },
            }
        }
    }
}

trait Identifier {
    fn is_identifier_starter(&self) -> bool;
    fn is_valid_identifier(&self) -> bool;
}

impl Identifier for char {
    fn is_identifier_starter(&self) -> bool {
        self.is_alphabetic() || matches!(self, '_' | '$')
    }

    fn is_valid_identifier(&self) -> bool {
        self.is_ascii_alphanumeric() || matches!(self, '_')
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // allowing dead code as this is simply a test helper
    #[derive(Debug)]
    #[allow(dead_code)]
    struct TokenSnapshot<'snap> {
        token: Token,
        content: &'snap str,
    }

    impl<'snap> TokenSnapshot<'snap> {
        fn with_content(token: Token, content: &'snap str) -> Self {
            Self { content, token }
        }
    }

    #[test]
    fn test_lexing_simple_document() {
        let source = [
            "<!DOCTYPE html>",
            "<html lang=\"en\">",
            "  <head id=\"head\">",
            "    <meta charset=\"UTF-8\">",
            "    <meta name=\"description\" content=\"{{ page.description }}\">",
            "",
            "    {{ if page.author }}",
            "    <title>{{ page.title }} - {{ page.author }}</title>",
            "    {{ else }}",
            "    <title>{{ page.title }}</title>",
            "    {{ end }}",
            "",
            "    {{ slot \"head\" }}",
            "  </head>",
            "  <body>",
            "    <main>",
            "      {{ slot \"page\" }}",
            "    </main>",
            "  </body>",
            "</html>",
        ]
        .join("\n");

        let source = Arc::new(source);
        let lexer = Lexer::new(source.clone());

        let tokens = lexer
            .map(|t| t.unwrap())
            .map(|token| TokenSnapshot::with_content(token, &source[token.loc().range()]))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(tokens);
    }

    #[test]
    fn test_lexing_language_features() {
        // this is purposefully nonsensical as it is supposed to test lexing of
        // language features.
        let source = [
            // if statement
            "{{ if page.author }}",
            "<title>{{ page.title }} - {{ page.author }}</title>",
            "{{ else }}",
            "<title>{{ page.title }}</title>",
            "{{ end }}",
            "",
            // for loop
            "{{ for tag in page.tags }}",
            "<span>{{ tag.name }}</span>",
            "{{ end }}",
            "",
            // slots
            "{{ slot \"head\" }}",
            "{{ slot \"page\" }}",
            "",
            // blocks
            "{{ block \"head\" }}",
            "{{ block \"page\" }}",
            // render
            "{{ render \"something.html\" }}",
            "{{ render \"another_thing.html\" }}",
            // extend
            "{{ extend \"template.html\" }}",
            // calling functions
            "{{ site.assets.get('asset_name.extension') }}",
            // variable
            "{{ name := page.title }}",
            // comparisons
            "{{ page.title == \"some title\" }}",
            "{{ page.title != \"some title\" }}",
            "{{ page.words > 1337 }}",
            "{{ page.words >= 1337 }}",
            "{{ page.words < 1337 }}",
            "{{ page.words <= 1337 }}",
            "{{ page.words != 1337 }}",
            // numbers
            "{{ 1337 }}",
            "{{ -1337 }}",
            "{{ 13.37 }}",
            "{{ -13.37 }}",
            "{{ -1337 }}",
            "{{ 13.37 }}",
            "{{ -13.37 }}",
            "{{ 1337 - -1337 }}",
            "{{ -1337 + -1337 }}",
            "{{ -1337+-1337 }}",
            "{{ -1337.1337+-13.33 }}",
        ]
        .join("\n");

        let source = Arc::new(source);
        let lexer = Lexer::new(source.clone());

        let tokens = lexer
            .map(|t| t.unwrap())
            .map(|token| TokenSnapshot::with_content(token, &source[token.loc().range()]))
            .collect::<Vec<_>>();

        insta::assert_debug_snapshot!(tokens);
    }
}
