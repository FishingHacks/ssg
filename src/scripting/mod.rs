use std::ops::{Add, AddAssign, Range};

use lexer::{Lexer, Token};

mod lexer;
mod parser;

#[derive(Debug, Clone, Copy)]
pub struct ByteOffset {
    start: usize,
    end: usize,
}

impl ByteOffset {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    pub fn range(self) -> Range<usize> {
        self.start..self.end
    }
}

impl Add for ByteOffset {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        Self {
            start: self.start.min(rhs.start),
            end: self.end.min(rhs.end),
        }
    }
}
impl AddAssign for ByteOffset {
    fn add_assign(&mut self, rhs: Self) {
        *self = *self + rhs;
    }
}

impl From<(usize, usize)> for ByteOffset {
    fn from((start, end): (usize, usize)) -> Self {
        debug_assert!(end > start);
        Self { start, end }
    }
}

#[derive(Debug)]
pub struct Ast {
    source: String,
    nodes: Vec<AstNode>,
}

#[derive(Debug)]
pub enum AstNode {
    // Does not include either quotes. Extend/shorten by 1 on either side to get the quotes too
    Html(ByteOffset),
    /// ` ┌───────────────┐ <- .0 `
    /// ` block "meow meow"       `
    /// `        └───────┘  <- .1 `
    Block(ByteOffset, ByteOffset),
    /// ` ┌───────────────┐ <- .0 `
    /// ` enter "meow meow"       `
    /// `        └───────┘  <- .1 `
    Enter(ByteOffset, ByteOffset, Box<[AstNode]>),
    /// ` ┌────────────────┐ <- .0 `
    /// ` extend "base.html"       `
    /// `         └───────┘  <- .1 `
    Extend(ByteOffset, ByteOffset),
    Slot(ByteOffset),
    If(ByteOffset, Box<Expr>, Box<[AstNode]>, Box<[AstNode]>),
    // 1.: Location of the statement, 2.: Location of the variable identfier
    For(ByteOffset, ByteOffset, Box<Expr>, Box<[AstNode]>),
    Render(ByteOffset, Box<Expr>, Box<[AstNode]>),
    // a := b + c
    /// ` ┌──────────────┐ <- .0 `
    /// ` my_var := <expr>       `
    /// ` └────┘           <- .1 `
    VarDef(ByteOffset, ByteOffset, Box<Expr>),
    Expr(Box<Expr>),
}

impl AstNode {
    fn loc(&self) -> ByteOffset {
        match self {
            AstNode::Html(byte_offset)
            | AstNode::Block(byte_offset, ..)
            | AstNode::Enter(byte_offset, ..)
            | AstNode::Extend(byte_offset, ..)
            | AstNode::Slot(byte_offset)
            | AstNode::If(byte_offset, ..)
            | AstNode::For(byte_offset, ..)
            | AstNode::Render(byte_offset, ..)
            | AstNode::VarDef(byte_offset, ..) => *byte_offset,
            AstNode::Expr(expr) => expr.loc(),
        }
    }
}

#[derive(Debug)]
pub enum Expr {
    /// ` ┌───────────────┐ <- .0 `
    /// ` "my fancy string"       `
    /// `  └─────────────┘  <- .1 `
    String(ByteOffset, ByteOffset),
    Var(ByteOffset),
    /// <expr _1>._2
    /// ` ┌──────────────┐ <- .0 `
    /// ` <expr>.something       `
    /// `        └───────┘ <- .2 `
    Index(ByteOffset, Box<Expr>, ByteOffset),
    // <expr _1>._2(<exprs _3>)
    /// ` ┌───────────────────┐ <- .0 `
    /// ` <expr>.something(...)       `
    /// `        └───────┘      <- .2 `
    FuncCall(ByteOffset, Box<Expr>, ByteOffset, Box<[Expr]>),
    // <expr _1> ?? <expr _2>
    Either(ByteOffset, Box<Expr>, Box<Expr>),
    // not <expr _1>
    Not(ByteOffset, Box<Expr>),
    // <expr _1> and <expr _2>
    And(ByteOffset, Box<Expr>, Box<Expr>),
    // <expr _1> or <expr _2>
    Or(ByteOffset, Box<Expr>, Box<Expr>),
    // <expr _1> == <expr _2>
    Equal(ByteOffset, Box<Expr>, Box<Expr>),
    // <expr _1> != <expr _2>
    NotEqual(ByteOffset, Box<Expr>, Box<Expr>),
    // <expr _1> + <expr _2>
    Concat(ByteOffset, Box<Expr>, Box<Expr>),
}

impl Expr {
    pub fn loc(&self) -> ByteOffset {
        match self {
            Expr::String(byte_offset, ..)
            | Expr::Var(byte_offset)
            | Expr::Index(byte_offset, ..)
            | Expr::FuncCall(byte_offset, ..)
            | Expr::Either(byte_offset, ..)
            | Expr::Not(byte_offset, ..)
            | Expr::And(byte_offset, ..)
            | Expr::Or(byte_offset, ..)
            | Expr::Equal(byte_offset, ..)
            | Expr::NotEqual(byte_offset, ..)
            | Expr::Concat(byte_offset, ..) => *byte_offset,
        }
    }
}

pub fn parse_template(template: String) -> miette::Result<Ast> {
    let mut ast = Ast {
        source: template,
        nodes: vec![],
    };

    let tokens = Lexer::new(&ast.source).lex()?;
    let mut parser = parser::Parser::new(&mut ast, &tokens);
    parser.parse_all().map_err(|v| miette::miette!("{v}"))?;

    Ok(ast)
}

pub fn print_token(src: &str, tok: &Token) {
    match tok {
        Token::Html(byte_offset) => println!("Html({:?})", &src[byte_offset.range()]),
        Token::ExprStart(_) => println!("ExprStart"),
        Token::ExprEnd(_) => println!("ExprEnd"),
        Token::Identifier(byte_offset) => println!("Ident({})", &src[byte_offset.range()]),
        Token::StringLiteral(byte_offset) => {
            println!("StringLit({:?})", &src[byte_offset.start + 1..byte_offset.end - 1])
        }
        Token::Keyword(byte_offset) => println!("Keyword({})", &src[byte_offset.range()]),
        Token::Symbol(c, _) => println!("Symbol({c:?})"),
    }
}

pub fn print_node(src: &str, node: &AstNode, indent: usize) {
    for _ in 0..indent {
        print!("    ");
    }
    match node {
        AstNode::Html(byte_offset) => println!("Html({:?})", &src[byte_offset.range()]),
        AstNode::Block(_, name) => println!("Block({:?})", &src[name.range()]),
        AstNode::Enter(_, name, children) => {
            println!("Enter({:?}, [", &src[name.range()]);
            children.iter().for_each(|v| print_node(src, v, indent + 1));
            for _ in 0..indent {
                print!("    ");
            }
            println!("])");
        }
        AstNode::Extend(_, file) => println!("Extend({:?})", &src[file.range()]),
        AstNode::Slot(_) => println!("Slot"),
        AstNode::If(_, expr, if_block, else_block) => {
            print!("If(");
            print_expr(src, expr);
            println!(", [");
            if_block.iter().for_each(|v| print_node(src, v, indent + 1));
            for _ in 0..indent {
                print!("    ");
            }
            println!("], [");
            else_block.iter().for_each(|v| print_node(src, v, indent + 1));
            for _ in 0..indent {
                print!("    ");
            }
            println!("])");
        }
        AstNode::For(_, var, expr, body) => {
            print!("For({}, ", &src[var.range()]);
            print_expr(src, expr);
            println!(", [");
            body.iter().for_each(|v| print_node(src, v, indent + 1));
            for _ in 0..indent {
                print!("    ");
            }
            println!("])");
        }
        AstNode::Render(_, expr, children) => {
            print!("Render(");
            print_expr(src, expr);
            println!(", [");
            children.iter().for_each(|v| print_node(src, v, indent + 1));
            for _ in 0..indent {
                print!("    ");
            }
            println!("])");
        }
        AstNode::VarDef(_, name, expr) => {
            print!("Define({}, ", &src[name.range()]);
            print_expr(src, expr);
            println!(")");
        }
        AstNode::Expr(expr) => {
            print!("Expr(");
            print_expr(src, expr);
            println!(")\n")
        }
    }
}

pub fn print_expr(src: &str, expr: &Expr) {
    match expr {
        Expr::String(_, range) => print!("{:?}", &src[range.range()]),
        Expr::Var(byte_offset) => print!("{}", &src[byte_offset.range()]),
        Expr::Index(_, lhs, rhs) => {
            print!("(index ");
            print_expr(src, lhs);
            print!(" {})", &src[rhs.range()]);
        }
        Expr::FuncCall(_, expr, name, exprs) => {
            print!("(call ");
            print_expr(src, expr);
            print!(", {}", &src[name.range()]);
            for expr in exprs {
                print!(", ");
                print_expr(src, expr);
            }
            print!(")");
        }
        Expr::Not(_, expr) => {
            print!("(not ");
            print_expr(src, expr);
            print!(")");
        }
        Expr::Either(_, lhs, rhs) => {
            print!("(either ");
            print_expr(src, lhs);
            print!(", ");
            print_expr(src, rhs);
            print!(")");
        }
        Expr::And(_, lhs, rhs) => {
            print!("(and ");
            print_expr(src, lhs);
            print!(", ");
            print_expr(src, rhs);
            print!(")");
        }
        Expr::Or(_, lhs, rhs) => {
            print!("(or ");
            print_expr(src, lhs);
            print!(", ");
            print_expr(src, rhs);
            print!(")");
        }
        Expr::Equal(_, lhs, rhs) => {
            print!("(equal ");
            print_expr(src, lhs);
            print!(", ");
            print_expr(src, rhs);
            print!(")");
        }
        Expr::NotEqual(_, lhs, rhs) => {
            print!("(notequal ");
            print_expr(src, lhs);
            print!(", ");
            print_expr(src, rhs);
            print!(")");
        }
        Expr::Concat(_, lhs, rhs) => {
            print!("(concat ");
            print_expr(src, lhs);
            print!(", ");
            print_expr(src, rhs);
            print!(")");
        }
    }
}
