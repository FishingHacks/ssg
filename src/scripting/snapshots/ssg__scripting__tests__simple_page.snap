---
source: src/scripting/mod.rs
expression: ast
---
Ast {
    source: "\n<!DOCTYPE html>\n<html lang=\"en\">\n    <head id=\"head\">\n        <meta charset=\"UTF-8\">\n\n        <meta name=\"description\" content=\"{{ page.description }}\">\n\n        <meta http-equiv=\"x-ua-compatible\" content=\"ie=edge\">\n        <meta name=\"viewport\" content=\"width=device-width, initial-scale=1, shrink-to-fit=no\">\n\n        {{ if page.author }}\n        <title>{{ page.title }} - {{ page.author }}</title>\n        {{ else }}\n        <title>{{ page.title }}</title>\n        {{ end }}\n\n        {{ slot \"head\" }}\n    </head>\n\n    <body>\n        <main>\n            {{ slot \"page\" }}\n        </main>\n    </body>\n</html>\n",
    nodes: [
        Html(
            ByteOffset {
                start: 0,
                end: 129,
            },
        ),
        Expr(
            Index(
                ByteOffset {
                    start: 132,
                    end: 148,
                },
                Var(
                    ByteOffset {
                        start: 132,
                        end: 136,
                    },
                ),
                ByteOffset {
                    start: 137,
                    end: 148,
                },
            ),
        ),
        Html(
            ByteOffset {
                start: 151,
                end: 321,
            },
        ),
        If(
            ByteOffset {
                start: 324,
                end: 475,
            },
            Index(
                ByteOffset {
                    start: 327,
                    end: 338,
                },
                Var(
                    ByteOffset {
                        start: 327,
                        end: 331,
                    },
                ),
                ByteOffset {
                    start: 332,
                    end: 338,
                },
            ),
            [
                Html(
                    ByteOffset {
                        start: 341,
                        end: 357,
                    },
                ),
                Expr(
                    Index(
                        ByteOffset {
                            start: 360,
                            end: 370,
                        },
                        Var(
                            ByteOffset {
                                start: 360,
                                end: 364,
                            },
                        ),
                        ByteOffset {
                            start: 365,
                            end: 370,
                        },
                    ),
                ),
                Html(
                    ByteOffset {
                        start: 373,
                        end: 376,
                    },
                ),
                Expr(
                    Index(
                        ByteOffset {
                            start: 379,
                            end: 390,
                        },
                        Var(
                            ByteOffset {
                                start: 379,
                                end: 383,
                            },
                        ),
                        ByteOffset {
                            start: 384,
                            end: 390,
                        },
                    ),
                ),
                Html(
                    ByteOffset {
                        start: 393,
                        end: 410,
                    },
                ),
            ],
            [
                Html(
                    ByteOffset {
                        start: 420,
                        end: 436,
                    },
                ),
                Expr(
                    Index(
                        ByteOffset {
                            start: 439,
                            end: 449,
                        },
                        Var(
                            ByteOffset {
                                start: 439,
                                end: 443,
                            },
                        ),
                        ByteOffset {
                            start: 444,
                            end: 449,
                        },
                    ),
                ),
                Html(
                    ByteOffset {
                        start: 452,
                        end: 469,
                    },
                ),
            ],
        ),
        Html(
            ByteOffset {
                start: 478,
                end: 488,
            },
        ),
        Slot(
            ByteOffset {
                start: 491,
                end: 495,
            },
        ),
        Expr(
            String(
                ByteOffset {
                    start: 497,
                    end: 501,
                },
                ByteOffset {
                    start: 498,
                    end: 500,
                },
            ),
        ),
        Html(
            ByteOffset {
                start: 505,
                end: 557,
            },
        ),
        Slot(
            ByteOffset {
                start: 560,
                end: 564,
            },
        ),
        Expr(
            String(
                ByteOffset {
                    start: 566,
                    end: 570,
                },
                ByteOffset {
                    start: 567,
                    end: 569,
                },
            ),
        ),
        Html(
            ByteOffset {
                start: 574,
                end: 611,
            },
        ),
    ],
}
